content/oci/oci.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/go-digest"
+	specs "github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/internal/graph"
+	reslv "oras.land/oras-go/v2/internal/resolver"
+)
+
+// ociImageIndexFile is the file name of the index from the OCI Image Layout Specification.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md#indexjson-file
+const ociImageIndexFile = "index.json"
+
+// Store implements `oras.Target`, and represents a content store
+// based on file system with the OCI-Image layout.
+// Reference: https://github.com/opencontainers/image-spec/blob/master/image-layout.md
+type Store struct {
+	storage  content.Storage
+	resolver *reslv.Memory
+	finder   *graph.Memory
+	index    *ocispec.Index
+
+	root      string
+	indexPath string
+}
+
+func New(root string) (*Store, error) {
+	store := &Store{
+		storage:   NewStorage(root),
+		resolver:  reslv.NewMemory(),
+		finder:    graph.NewMemory(),
+		root:      root,
+		indexPath: filepath.Join(root, ociImageIndexFile),
+	}
+
+	if err := store.ensureOCILayoutFile(); err != nil {
+		return nil, err
+	}
+
+	if err := store.loadIndex(); err != nil {
+		return nil, err
+	}
+
+	return store, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (s *Store) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	return s.storage.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (s *Store) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	if err := s.storage.Push(ctx, expected, reader); err != nil {
+		return err
+	}
+
+	return s.finder.Index(ctx, s.storage, expected)
+}
+
+// Exists returns true if the described content exists.
+func (s *Store) Exists(ctx context.Context, target ocispec.Descriptor) (bool, error) {
+	return s.storage.Exists(ctx, target)
+}
+
+// Tag tags a descriptor with a reference string.
```

Comment:
I noticed that for digest-referenced manifest, `containerd`  stores the digest in the `org.opencontainers.image.ref.name` annotation as `"org.opencontainers.image.ref.name":  "@sha256:123abc". I'm going for this way.

Reply:
What if the user invokes `s.Tag(ctx, desc, "localhost:5000/myrepo:mytag")` since it is a common scenario in `Copy()`. Will you parse it and only store `mytag` in `org.opencontainers.image.ref.name`?

Reply:
or consider `localhost:5000/myrepo:mytag` as an invalid reference?